src/ui/public/notify/banners/banners.js

```diff
@@ -0,0 +1,106 @@
+/**
+ * Banners represents a prioritized list of displayed components.
+ */
+export class Banners {
+
+  constructor() {
+    // sorted in descending order (100, 99, 98...) so that higher priorities are in front
+    this.list = [];
+    this.uniqueId = 0;
+    this.onChangeCallback = null;
+  }
+
+  _changed = () => {
+    if (this.onChangeCallback) {
+      this.onChangeCallback();
+    }
+  }
+
+  _replaceIfExists = bannerDetails => {
+    const index = this.list.findIndex(details => details.id === bannerDetails.id);
+
+    // we found the banner
+    if (index !== -1) {
+      // if its priority didn't change, then we don't need to worry about order
+      if (this.list[index].priority === bannerDetails.priority) {
+        this.list[index] = bannerDetails;
+
+        return true;
+      }
+
+      // the priority did change, so we just remove it and let 'set' do its magic
+      this.list.splice(index, 1);
+    }
+
+    return false;
+  }
+
+  /**
+   * Set the {@code callback} to invoke whenever changes are made to the banner list.
+   *
+   * Use {@code null} or {@code undefined} to unset it.
+   *
+   * @param {Function} callback The callback to use.
+   */
+  onChange = callback => {
+    this.onChangeCallback = callback;
+  }
+
+  /**
+   * Add a new banner or replace an existing one with the optional {@code id}.
+   *
+   * @param {Object} component The React component to display.
+   * @param {String} id The optional ID of the banner to replace.
+   * @param {Number} priority The optional priority order to display this banner. Higher priority values are shown first.
+   * @return {String} The supplied ID if {@code id} was not undefined. This value can be used to remove the banner.
+   */
+  add = ({ component, id, priority = 0 }) => {
```

Comment:
Can we use a default here for `id`?
```js
({ component, id = `banner-${++this.uniqueId}`, priority = 0 })
```

Reply:
Is the `banner` prefix useful? Unless there's a use case I'm missing, it seems like needless complexity (however small) so maybe we drop it?

Reply:
I was thinking about someone stuffing its value in an `id` attribute of an HTML tag, which cannot start with numbers if I remember correctly.